Feat : 개인정보 설정 모달창 제작

hasol11 opened the pull request:
## 📝작업 내용

![image](https://github.com/user-attachments/assets/a15858d2-1996-458e-8522-7773cad4c040)

- 이 모달창은 홈페이지에서 회원가입 후 학번과 학과 등 개인정보를 입력할 수 있습니다.
- 각 요소(InfoMessage, UserInfoForm, Button)는 개별 래퍼로 감싸져 있으며, 이를 통해 간격을 조정했습니다. 추가로 모달창이 밑에서 위로 올라오는 애니메이션을 추가했습니다.

### 추후 추가해야 하는 코드
- 입력값이 주어진 후 버튼을 누르면 데이터를 보내고 모달창이 닫히도록 기능을 추가할 예정입니다.

### 참고
현재 버튼 컴포넌트가 PR이 완료되지 않은 상태라 임시로 지연님이 만들어주신 버튼 컴포넌트의 내용을 옮겨와서 사용했습니다. 혹시 발생할 수 있는 이슈에 대비해 말씀드립니다!

## 💬리뷰 요구사항
혹시 잘못된 부분이나 수정하면 좋을 부분이 있다면 말씀해주세요!

hasol11 committed:
feat: 개인정보 설정 모달창 제작

hasol11 committed:
refactor:buttonWrapper로 간격 조정

hasol11 committed:
Update UserInfoModal.tsx

refactor: 주석 정리

hanheel committed:
Design : info-message의 디자인 변경

1. 아이콘과 메세지를 상단으로 정렬
2. 텍스트의 line-height를 조절하여 피그마와 유사하게 출력

hanheel committed:
refactor : 컴포넌트 분리 및 파일 구조 변경

1. UserInfoModal -> Modal 로 변경하여 재활용성을 높임
2. contents에 Modal에 들어갈 컨텐츠를 모아 유지보수를 편하게 만듦
3. form 을 따로 분리, 추후 로직을 작성하는 것을 편하게 만듦

hanheel commented:
keyframe으로 애니메이션 잘 설정해 주셨네요! 수고 많으셨습니다!! 항상 느끼지만 어느 컴포넌트인지 아주 명확하게 잘 드러나서 추가적인 주석도 필요없고, 처음 코드를 읽어도 금방 구조를 알 수 있는 것 같습니다. 

### InputMessage 디자인 변경
만들어주신 모달을 제대로 확인하기 위해 따로 브랜치를 파서 수정하지 않고 이 브랜치에서 InfoMessage 컴포넌트를 함께 수정했습니다. 피드백 감사해요!

몇가지 앞으로 프로젝트를 진행하는 데에 있어서 리팩토링해야 할 부분이 조금 걸려서
**1. 몇가지 제가 직접 리팩토링한 부분**
**2. 추가적으로 부탁드릴 사안**
이렇게 제안드리려고 합니다!
(이번에 직접 리팩토링을 진행한 건, 저도 코드리뷰하면서 남의 코드를 수정하고 읽어보는 경험이 필요해서 지윤님 코드를 잠시 빌렸어요!)

**참고로 버튼 컴포넌트는 완성되지 않아, 길게 늘어지게 보일겁니다**

## 직접 리팩토링한 부분
### 컴포넌트의 재활용성
컴포넌트는 사실 **재활용**을 할 수있다는 특징이 가장 큰 메리트인 것 같습니다. 그래서 다른 페이지에서 사용되는 부분이 있다면, 쪼개어서 따로 만들어두는 게 가장 효율적인 방식이라고 생각해요. 지금 만들어주신 컴포넌트내에 모달과 모달 내부 컨텐츠가 함께 들어가 있는데, 피그마를 보시면 모달이라는 구조를 사용해야 하는 페이지가 꽤 있다는 걸 알 수 있어요. 따라서 모달에 '컨텐츠 컴포넌트' 자체를 전달하는 방식을 썼을 때, 모달 컴포넌트를 지금보다 더 재활용할 수 있다는 게 눈에 보이실 겁니다! ContentBubble (말풍선) 컴포넌트도 동일한 방식으로 제작해서, 말풍선 레이아웃을 여러 번 재활용 할 수 있도록 했어요!
(+Input 도 따로 분리했습니다!)

사용방법은 아래와 같습니다
```javascript
function App() {
  return <Modal contents={<InitialInfo />} />;
}
```

### 탐색의 용이성
지금 제작해주신 컴포넌트에 '폼'이 함께 포함되어 있는데, 이럴경우 해당 폼의 기능을 개발하거나 리팩토링하기 위해 찾을 때 '모달 > 폼?' 이라는 경로가 직관적이지 못해 찾는데 시간이 걸릴 수 있어요. (기능 개발에 들어가면 알겠지만 폼에는 정말 많고..긴..코드가 들어가서 따로 빼는 게 더 낫기도 합니다) 그래서 컴포넌트 폴더에 form 폴더를 따로 파서 폼 부분만 떼어 옮겨줬습니다!

**한 가지 궁금한 건, 저희...input과 form은 다른 분이 맡으시지 않았나요??? 제가 잘못 전달드렸다면 죄송합니다😅**

### max-width, max-height, min-width, min-height 속성을 활용할 것
모달을 감싸고 있는 width, height를 px로 설정해주셨습니다. 그런데 이렇게 안에 컨텐츠에 따라 크기가 달라지는 레이아웃은 "최소 크기"를 둬서 디자인의 통일성을 두되, 혹시라도 컨텐츠가 너무 커서 크기가 부족한 것을 대비해 padding값을 줘 너비와 높이가 변하게 한다면 더 유동적인 컴포넌트를 만들 수 있어요!
(컨텐츠가 적다면 최소 크기가 적용되는데, 최소크기보다 크다면 맞춰서 커지도록 구현하는 것을 설명드리고 싶었습니다!)

### margin으로 컴포넌트 배치하는 것을 최소화하면 좋을 것 같아요
제가 초반에도 flex나 grid 배치를 추천드렸던 이유는, margin으로 컴포넌트 사이를 배치하게 되면 예상치 못한 동작이 생길 수 있기 때문입니다!
얘를 들어, 10개의 컴포넌트를 감싸는 wrap 컴포넌트를 만들고 gap을 설정했을 때, 특정 컴포넌트에 margin이 설정돼있었다면 같은 간격으로 떨어지는 게 아닌 특정 컴포넌트만 저 멀리에 위치하는 현상이 발생할 수 있어요. 근데 10개의 컴포넌트를 일일이 살펴보면서 예상치 못한 동작을 찾기엔 시간이 너무 오래걸리겠죠.... (사실 제 경험입니다) 이 프로젝트에서는 flex를 이용해 배치해보면 어떨까요!

### 최종 파일 구조
```c
components
ㄴform
  ㄴUserInfoForm.tsx // 이름 , 학번 입력하는 폼 분리
ㄴmodal
  ㄴcontents
    ㄴUserInfo.tsx  // 모달에 들어갈 컨텐츠
ㄴModal.tsx // 모달 레이아웃
```

## 추가적으로 부탁드릴 사항
### 하드코딩을 줄이면 유지보수가 편해집니다!
```javascript
     content={`회원가입 후 최초 1회, 원활한 서비스 이용을 위해
학번, 학과, 이름 등의 개인정보 설정이 필요합니다`}
```
해당 부분은 InfoMessage의 message 부분에 넣는 텍스트입니다. jsx(html처럼 적는 그 부분이요!) 부분에 직접적으로 하드코딩해주셨는데, jsx 부분은 길어지면 길어질수록 가독성이 정말 떨어지더라구요. 또한, 나중에 텍스트가 바뀌거나 수정할 때를 대비해 변수로 빼서 관리하면 더 좋을 것 같다는 생각을 했어요! 이 부분 수정하면 어떨까요?

### 모달 기능 개발 (꼭 오늘까지 하지 않아도 괜찮습니다)
저희 사이트의 모달은 특정 버튼을 누르거나 화면에 접근하면 창이 뜨고, 또 모달 내의 버튼을 누르면 창이 꺼지는 방식을 가지고 있습니다. 이 기능이 모달의 꽃이라, 여기까지 한 번 부탁을 드려보려고 해요

1. 화면 내에 "임시 버튼"을 생성한다
2.  state를 이용해서, 화면의 버튼을 누르면 모달이 켜지고, 모달 내의 버튼을 누르면 꺼지도록 구현한다
3. 해당 로직만 분리해서 js  파일로 만든다!

3번이 좀 까다로울 수 있는데, 그러면 2번까지만 하고 같이 로직 분리를 해도 좋을 것 같아요!

hanheel committed:
Design: max, min width 를 사용하여 유동성을 높인다

hanheel committed:
refactor : 재활용성을 고려해 input컴포넌트를 분리한다

1. type - text, textarea 등 다양한 inputType을 입력받는다
2. label - input 옆에 적힐 label 값

hanheel committed:
Design : reset.css를 통해 input의 초기값을 변경한다

기존) outline으로 인해 테두리 변경이 보이지 않음 / border 색상을 사용하는 컴포넌트 마다 변경해야 함
변경 후) reset.css에 input태그를 사용할 경우, 테두리가 없고 border는 기본 light-primary 컬러로 설정되도록 변경

hasol11 committed:
refactor: infoMessage 텍스트 변수로 변환

hasol11 committed:
refactor: placeholder 내용 수정

hasol11 committed:
refactor: infoMessageText 변수명 변경

hasol11 committed:
refactor : 모달 열고 닫는 기능 추가

hasol11 committed:
refactor: 모달창 열고 닫는 기능 수정 (추가 수정예정)

hanheel commented:
추가적으로 두 가지만 요청드립니다!

1. closeModal -> handleModal / openModal -> handleModal ("함수"를 전달한다는 느낌이 들 수 있도록 명칭 통일)
2. 해당 모달창을 열고 닫는 로직도 재활용이 필요합니다! : 커스텀 훅으로 모달을 열고 닫는 로직을 분리하여 사용할 수 있도록 함
